R/PipeOpLearnerCV.R

```diff
@@ -159,53 +165,71 @@ PipeOpLearnerCV = R6Class("PipeOpLearnerCV",
-      if (pv$method != "insample") {
-        rdesc = mlr_resamplings$get(pv$method)
-        if (pv$method == "cv") rdesc$param_set$values = list(folds = pv$folds)
-        rr = resample(task, private$.learner, rdesc)
-        prds = as.data.table(rr$prediction(predict_sets = "test"))
-      } else {
-        prds = as.data.table(private$.learner$predict(task))
-      }
+      # compute resampled predictions
+      rr = resample(task, private$.learner, private$.resampling)
+      prds = as.data.table(rr$prediction(predict_sets = "test"))
 
-      private$pred_to_task(prds, task)
+      private$.pred_to_task(prds, task)
     },
 
     .predict_task = function(task) {
       on.exit({private$.learner$state = NULL})
       private$.learner$state = self$state
-      prediction = as.data.table(private$.learner$predict(task))
-      private$pred_to_task(prediction, task)
+      prds = as.data.table(private$.learner$predict(task))
+      private$.pred_to_task(prds, task)
     },
 
-    pred_to_task = function(prds, task) {
-      if (!is.null(prds$truth)) prds[, truth := NULL]
-      if (!self$param_set$values$resampling.keep_response && self$learner$predict_type == "prob") {
+    .pred_to_task = function(prds, task) {
+      if (!self$param_set$values$keep_response && self$learner$predict_type == "prob") {
         prds[, response := NULL]
       }
-      renaming = setdiff(colnames(prds), c("row_id", "row_ids"))
-      setnames(prds, renaming, sprintf("%s.%s", self$id, renaming))
+      renaming = setdiff(colnames(prds), c("row_ids", "truth"))
+      setnames(prds, old = renaming, new = sprintf("%s.%s", self$id, renaming))
+      setnames(prds, old = "truth", new = task$target_names)
+      row_reference = paste0("pre.", self$id)
+      while (row_reference %in% task$col_info$id) {
+        row_reference = paste0(row_reference, ".")
+      }
+      setnames(prds, old = "row_ids", new = row_reference)
 
-      # This can be simplified for mlr3 >= 0.11.0;
-      # will be always "row_ids"
-      row_id_col = intersect(colnames(prds), c("row_id", "row_ids"))
-      setnames(prds, old = row_id_col, new = task$backend$primary_key)
-      task$select(character(0))$cbind(prds)
+      # the following is needed to pertain correct row ids in the case of e.g. cv
+      # here we do not necessarily apply PipeOpAggregate later
+      backend = if (identical(sort(prds[[row_reference]]), sort(task$row_ids))) {
+        set(prds, j = task$backend$primary_key, value = prds[[row_reference]])
+        as_data_backend(prds, primary_key = task$backend$primary_key)
+      } else {
+        as_data_backend(prds)
+      }
+
+      # get task_type from mlr_reflections and call constructor
+      constructor = get(mlr_reflections$task_types[["task"]][chmatch(task$task_type, table = mlr_reflections$task_types[["type"]], nomatch = 0L)][[1L]])
+      newtask = invoke(constructor$new, id = task$id, backend = backend, target = task$target_names, .args = task$extra_args)
```

Comment:
Note to @mb706, this needs to be brought in accord with PipeOpTaskPreproc's `affect_columns`.

Reply:
But it should be the PipeOpTaskPreproc's responsibility to keep all col-roles (that are not disabled by affect_columns), so in particular should respect weights etc.

Maybe use [this](https://github.com/mlr-org/miesmuschel/blob/188b2ad70e6e563ab3ee018fd8e94ceacbd5bc89/R/PipeOpDensitySplit.R#L111-L115) for inspiration.

Reply:
Also the previous-id-column should really be a different col-role to avoid accidentally training on the ID.